### PR TITLE
Scanner status updating endpoint

### PR DESCRIPTION
**Pull request by skymandr**
Scanner status endpoint. When posting a status, the status of the scanner is updated (actually added). If scanner doesn't exist it is created with a semi-random name. Get endpoint returns latest scanner status if available.

There are probably some tests missing.

### CODE REVIEW

File: scanomatic/io/scanning_store.py [hunk 1]
```diff
@@ -12,7 +13,11 @@ class ScanJobUnknownError(ValueError):
 
 Scanner = namedtuple(
     'Scanner',
-    ['name', 'power', 'owner', 'identifier']
+    ['name', 'power', 'owner', 'identifier', 'status']
+)
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: I'm not sure the list of statuses should be included in the `Scanner` object. It might become quite long and it's not clear  to me why that's useful. I'd keep them separated for now.

File: scanomatic/io/scanning_store.py [hunk 2]
```diff
@@ -12,7 +13,11 @@ class ScanJobUnknownError(ValueError):
...
+ScannerStatus = namedtuple(
+    'ScannerStatus',
+    ['job', 'time', 'status']
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: I'm not sure what `ScannerStatus.status` means?

File: scanomatic/io/scanning_store.py [hunk 3]
```diff
@@ -88,3 +94,15 @@ def get_current_scanjob(self, scanner_id, timepoint):
 
     def has_current_scanjob(self, scanner_id, timepoint):
         return self.get_current_scanjob(scanner_id, timepoint) is not None
+
+    def update_status(self, scanner_id, status):
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: s/update/add/

File: scanomatic/io/scanning_store.py [hunk 4]
```diff
@@ -88,3 +94,15 @@ def get_current_scanjob(self, scanner_id, timepoint):
...
+        now = datetime.utcnow()
+        job = self.get_current_scanjob(scanner_id, now)
+        self.get_scanner(scanner_id).status.append(
+            ScannerStatus(job, now, status))
+
+    def get_scanner_status(self, scanner_id):
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: s/status/statuses/

File: scanomatic/io/scanning_store.py [hunk 5]
```diff
@@ -88,3 +94,15 @@ def get_current_scanjob(self, scanner_id, timepoint):
 
     def has_current_scanjob(self, scanner_id, timepoint):
         return self.get_current_scanjob(scanner_id, timepoint) is not None
+
+    def update_status(self, scanner_id, status):
+        now = datetime.utcnow()
+        job = self.get_current_scanjob(scanner_id, now)
+        self.get_scanner(scanner_id).status.append(
+            ScannerStatus(job, now, status))
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: - I don't think creating the `ScannerStatus` instance should be there: this module is for storing/retrieving entities that are create outside. (This is also how all other method work).
- the job should be reported by the scanner

File: scanomatic/ui_server/scanners_api.py [hunk 6]
```diff
@@ -33,6 +33,31 @@ def scanner_get(scanner):
     )
 
 
+@blueprint.route("/<scanner>/status", methods=['POST'])
```
Comment (gdetrez) [inline scanomatic/ui_server/scanners_api.py]: If you implement the endpoint as a collection, I think it should be pluralized.

File: scanomatic/ui_server/scanners_api.py [hunk 7]
```diff
@@ -33,6 +33,31 @@ def scanner_get(scanner):
...
+    scanning_store = current_app.config['scanning_store']
+    if not scanning_store.has_scanner(scanner):
+        # TODO: should create non-existent
+        return json_abort(
+            NOT_FOUND, reason="Scanner '{}' unknown".format(
+                scanner)
+        )
+
+    status = request.get_json(silent=True, force=True)
+    scanning_store.scanner_status_update(scanner, status["message"])
```
Comment (gdetrez) [inline scanomatic/ui_server/scanners_api.py]: There should probably be some input validation here.

File: scanomatic/ui_server/scanners_api.py (hunk 6 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: I don't understand what you mean. Care to clarify?

File: scanomatic/io/scanning_store.py (hunk 4 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: or statii? ;)

File: scanomatic/io/scanning_store.py (hunk 3 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: ok

File: scanomatic/io/scanning_store.py (hunk 5 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: ok

File: scanomatic/ui_server/scanners_api.py (hunk 7 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: probably, yes.

Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: provisionally done

File: scanomatic/io/scanning_store.py (hunk 5 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py (hunk 3 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py (hunk 2 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: changed to `ScannerStatus.message`

File: scanomatic/io/scanning_store.py (hunk 1 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: I'll see what I can do about that.

File: scanomatic/io/scanning_store.py (hunk 4 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py (hunk 1 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py (hunk 2 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py [hunk 8]
```diff
@@ -14,6 +14,10 @@ class ScanJobUnknownError(ValueError):
     'Scanner',
     ['name', 'power', 'owner', 'identifier']
 )
+ScannerStatus = namedtuple(
+    'ScannerStatus',
+    ['job', 'time', 'message']
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: - what is message ?
- Maybe specify s/time/server_time/ (or recetption time or similar) ? I find it a bit ambiguous otherwise. We might also want to have both sending time on the scanner and reception time on the server to monitor the scanner clocks (I'm not asking that you add this here though).

File: scanomatic/ui_server/scanners_api.py [hunk 9]
```diff
@@ -33,6 +34,43 @@ def scanner_get(scanner):
...
+        return json_abort(
+            NOT_FOUND,
+            reason="Scanner '{}' unknown".format(scanner)
+        )
+
+    status = request.get_json(silent=True, force=True)
+    try:
+        scanning_store.add_scanner_status(
+            scanner, ScannerStatus(
+                status["job"], datetime.utcnow(), status["message"]))
```
Comment (gdetrez) [inline scanomatic/ui_server/scanners_api.py]: Use `datetime.now(pytz.utc)` instead of `utcnow` for a timezone-aware datetime.

File: tests/unit/ui_server/test_scanners_api.py [hunk 10]
```diff
@@ -41,24 +42,55 @@ class TestScannerStatus:
...
+        assert response.status_code == OK
         assert response.json == self.SCANNER
 
     def test_get_unknown_scanner(self, test_app):
         response = test_app.get(self.URI + "/Unknown")
-        response.status_code == NOT_FOUND
+        assert response.status_code == NOT_FOUND
         assert response.json['reason'] == "Scanner 'Unknown' unknown"
 
+    def test_get_and_add_scanner_status(self, test_app):
```
Comment (gdetrez) [inline tests/unit/ui_server/test_scanners_api.py]: This should be two tests

File: tests/unit/ui_server/test_scanners_api.py [hunk 11]
```diff
@@ -41,24 +42,55 @@ class TestScannerStatus:
...
+            data=json.dumps({"job": "foo", "message": "bar"}),
+            headers={'Content-Type': 'application/json'}
+        )
+        assert response.status_code == OK
+
+        response = test_app.get(
+            self.URI + "/9a8486a6f9cb11e7ac660050b68338ac/status")
+        assert response.status_code == OK
+        assert response.json["job"] == "foo"
+        assert response.json["message"] == "bar"
```
Comment (gdetrez) [inline tests/unit/ui_server/test_scanners_api.py]: You can use [freezegun](https://github.com/spulec/freezegun) to check the time field.

File: scanomatic/ui_server/scanners_api.py (hunk 6 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: ping @gdetrez

File: scanomatic/io/scanning_store.py (hunk 8 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: - message is whatever the scanner says (currently an empty string, so perhaps redundant).
- sure, `server_time` sits well with me.

File: scanomatic/ui_server/scanners_api.py (hunk 9 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: ok

File: tests/unit/ui_server/test_scanners_api.py (hunk 11 above)
Comment (skymandr) [reply tests/unit/ui_server/test_scanners_api.py]: cool!

File: scanomatic/ui_server/scanners_api.py (hunk 9 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: done

File: tests/unit/ui_server/test_scanners_api.py (hunk 10 above)
Comment (skymandr) [reply tests/unit/ui_server/test_scanners_api.py]: suggestions welcome

File: tests/unit/ui_server/test_scanners_api.py (hunk 11 above)
Comment (skymandr) [reply tests/unit/ui_server/test_scanners_api.py]: Done

File: scanomatic/io/scanning_store.py (hunk 8 above)
Comment (gdetrez) [reply scanomatic/io/scanning_store.py]: I didn't know scanners had something to say :robot::speech_balloon:

File: scanomatic/ui_server/scanners_api.py (hunk 6 above)
Comment (gdetrez) [reply scanomatic/ui_server/scanners_api.py]: If it respond with a list of statuses on GET, as opposed to a single status object, it should be called `/<scannner>/statuses`

File: scanomatic/io/scanning_store.py [hunk 12]
```diff
@@ -14,6 +14,10 @@ class ScanJobUnknownError(ValueError):
     'Scanner',
     ['name', 'power', 'owner', 'identifier']
 )
+ScannerStatus = namedtuple(
```
Comment (local-minimum) [inline scanomatic/io/scanning_store.py]: Adding a `ScannerStatus` should mean that the `owner` of the `Scanner` is redundant and thus needs removing.

Comment (local-minimum) [reply scanomatic/io/scanning_store.py]: Function on line 42 needs updating.

File: scanomatic/ui_server/scanners_api.py [hunk 13]
```diff
@@ -34,6 +35,54 @@ def scanner_get(scanner):
     )
 
 
+@blueprint.route("/<scanner>/status", methods=['POST'])
+def scanner_status_update(scanner):
+    scanning_store = current_app.config['scanning_store']
+    if not scanning_store.has_scanner(scanner):
+        # TODO: should create non-existent
```
Comment (local-minimum) [inline scanomatic/ui_server/scanners_api.py]: :fire:

File: scanomatic/ui_server/scanners_api.py [hunk 14]
```diff
@@ -34,6 +35,54 @@ def scanner_get(scanner):
...
+        return json_abort(
+            NOT_FOUND,
+            reason="Scanner '{}' unknown".format(scanner)
+        )
+
+    status = request.get_json(silent=True, force=True)
```
Comment (local-minimum) [inline scanomatic/ui_server/scanners_api.py]: skip silent and force.

File: scanomatic/ui_server/scanners_api.py (hunk 13 above)
Comment (local-minimum) [reply scanomatic/ui_server/scanners_api.py]: It should not abort it should add the scanner. Should be quite small thing to add. Since we removed registering scanners it needs to be part of this issue/task/pr.

File: scanomatic/ui_server/scanners_api.py [hunk 15]
```diff
@@ -34,6 +35,54 @@ def scanner_get(scanner):
...
+
+
+@blueprint.route("/<scanner>/status", methods=['GET'])
+def scanner_status_get(scanner):
+    scanning_store = current_app.config['scanning_store']
+    if scanning_store.has_scanner(scanner):
+        status = scanning_store.get_latest_scanner_status(scanner)
+        if status is None:
+            status = ScannerStatus(None, None, None)
+        return jsonify(**status._asdict())
```
Comment (local-minimum) [inline scanomatic/ui_server/scanners_api.py]: Return messages from the api in json-format should follow javascript syntax so add a function in the serializer module that converts the keys to camel case where needed.

Comment (local-minimum) [reply scanomatic/ui_server/scanners_api.py]: This applies to getting the list too.

File: scanomatic/ui_server/scanners_api.py [hunk 16]
```diff
@@ -34,6 +35,54 @@ def scanner_get(scanner):
...
+    return json_abort(
+        NOT_FOUND, reason="Scanner '{}' unknown".format(scanner)
+    )
+
+
+@blueprint.route("/<scanner>/statuses", methods=['GET'])
```
Comment (local-minimum) [inline scanomatic/ui_server/scanners_api.py]: feels like the endpoint should either be `/<scanner>/status/all` or `/<scanner>/status` but passing a setting with the request to filter what status one wants.

File: scanomatic/ui_server/scanners_api.py [hunk 17]
```diff
@@ -34,6 +35,54 @@ def scanner_get(scanner):
...
+def scanner_statuses_get(scanner):
+    scanning_store = current_app.config['scanning_store']
+    if scanning_store.has_scanner(scanner):
+        statuses = scanning_store.get_scanner_status_list(scanner)
+        return jsonify({"status_list": statuses})
```
Comment (local-minimum) [inline scanomatic/ui_server/scanners_api.py]: should just return the list (with the correctly formatted statuses), not put it on a key.

File: tests/unit/ui_server/test_scanners_api.py [hunk 18]
```diff
@@ -37,20 +43,104 @@ class TestScannerStatus:
 
     def test_get_all_implicit(self, test_app):
         response = test_app.get(self.URI)
-        response.status_code == OK
+        assert response.status_code == OK
```
Comment (local-minimum) [inline tests/unit/ui_server/test_scanners_api.py]: You'll have fun conflicting with me here.

File: tests/unit/ui_server/test_scanners_api.py [hunk 19]
```diff
@@ -37,20 +43,104 @@ class TestScannerStatus:
...
+                data=json.dumps({"job": "foo", "message": "bar"}),
+                headers={'Content-Type': 'application/json'}
+            )
+            assert response.status_code == OK
+
+            response = test_app.get(
+                self.URI + "/9a8486a6f9cb11e7ac660050b68338ac/status")
+            assert response.status_code == OK
+            assert response.json["job"] == "foo"
+            assert response.json["server_time"] == "1985-10-26T01:20:00Z"
```
Comment (local-minimum) [inline tests/unit/ui_server/test_scanners_api.py]: Should be `serverTime`

File: tests/unit/ui_server/test_scanners_api.py [hunk 20]
```diff
@@ -37,20 +43,104 @@ class TestScannerStatus:
...
+        response = test_app.get(self.URI + "/42/status")
+        assert response.status_code == NOT_FOUND
+
+    def test_add_bad_scanner_status_fails(self, test_app):
+        response = test_app.post(
+            self.URI + "/9a8486a6f9cb11e7ac660050b68338ac/status",
+            data=json.dumps({"foo": "foo", "bar": "bar"}),
+            headers={'Content-Type': 'application/json'}
+        )
+        assert response.status_code == BAD_REQUEST
```
Comment (local-minimum) [inline tests/unit/ui_server/test_scanners_api.py]: Missing test for posting status for scanner not yet known.

File: tests/unit/ui_server/test_scanners_api.py [hunk 21]
```diff
@@ -37,20 +43,104 @@ class TestScannerStatus:
...
+
+
+class TestGetScannerJob(object):
+    URI = '/api/scanners/xxxx/job'
+
+    @pytest.fixture
+    def fakedb(self):
+        return mock.MagicMock()
```
Comment (local-minimum) [inline tests/unit/ui_server/test_scanners_api.py]: why do you need to mock the db? Should be easy to instantiate the db

Comment (gdetrez) [reply tests/unit/ui_server/test_scanners_api.py]: That's the right way to write unit tests

Comment (gdetrez) [reply tests/unit/ui_server/test_scanners_api.py]: Actually, this is not new code. It should not be there...

File: scanomatic/io/scanning_store.py (hunk 12 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/ui_server/scanners_api.py (hunk 14 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: done

File: tests/unit/ui_server/test_scanners_api.py (hunk 18 above)
Comment (skymandr) [reply tests/unit/ui_server/test_scanners_api.py]: :)

File: scanomatic/ui_server/scanners_api.py (hunk 15 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: do you mean in `jsonify`?

File: tests/unit/ui_server/test_scanners_api.py (hunk 19 above)
Comment (skymandr) [reply tests/unit/ui_server/test_scanners_api.py]: done

File: scanomatic/ui_server/scanners_api.py (hunk 15 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: found it, done

File: scanomatic/io/scanning_store.py [hunk 22]
```diff
@@ -88,3 +92,21 @@ def get_current_scanjob(self, scanner_id, timepoint):
...
+    def get_latest_scanner_status(self, scanner_id):
+        try:
+            return self.get_scanner_status_list(scanner_id)[-1]
+        except IndexError:
+            return None
+
+    def add_scanner_status(self, scanner_id, status):
+        self.get_scanner_status_list(scanner_id).append(status)
+
+    def is_free(self, scanner_id):
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: I think it would be safer to look at the scanjobs directly to figure out if a scanner is free. (i.e. use `has_current_scanjob`)

Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: IMO it would be safer to look at the scanjobs directly. I.e. use `has_current_scanjob`.

Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py [hunk 23]
```diff
@@ -13,7 +17,11 @@ class ScanJobUnknownError(ValueError):
 
 Scanner = namedtuple(
     'Scanner',
-    ['name', 'power', 'owner', 'identifier']
+    ['name', 'power', 'identifier']
```
Comment (local-minimum) [inline scanomatic/io/scanning_store.py]: `power` should be removed with this PR since it is a status

File: tests/unit/ui_server/test_scanners_api.py (hunk 20 above)
Comment (skymandr) [reply tests/unit/ui_server/test_scanners_api.py]: done

File: scanomatic/io/scanning_store.py [hunk 24]
```diff
@@ -24,20 +32,28 @@ def __init__(self):
...
             self._scanners = {}
+        self._scanner_statuses = {scanner: [] for scanner in self._scanners}
         self._scanjobs = {}
 
+    def add_scanner(self, scanner_id, name=None):
+        if self.has_scanner(scanner_id):
+            raise ValueError("Cannot add duplicate scanner!")
+
+        if name is None:
+            name = get_generic_name(scanner_id)
```
Comment (local-minimum) [inline scanomatic/io/scanning_store.py]: Duplicate names are also not allowed so this has to verified somehow both when generating and not

File: scanomatic/io/scanning_store.py (hunk 8 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: fixed

File: scanomatic/ui_server/serialization.py [hunk 25]
```diff
@@ -14,3 +14,15 @@ def job2json(job):
         assert is_utc(job.start_time)
         obj['startTime'] = job.start_time.strftime('%Y-%m-%dT%H:%M:%SZ')
     return obj
+
+
+def status2json(status):
+    obj = {
+        'job': status.job,
+    }
+    if status.server_time is None:
```
Comment (local-minimum) [inline scanomatic/ui_server/serialization.py]: Don't put `serverTime` in if it doesn't exist (to match behaviour of the `job2json`.

File: scanomatic/ui_server/scanners_api.py (hunk 13 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: done

File: scanomatic/ui_server/scanners_api.py (hunk 16 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: removed

File: scanomatic/ui_server/scanners_api.py (hunk 17 above)
Comment (skymandr) [reply scanomatic/ui_server/scanners_api.py]: removed

File: tests/unit/ui_server/test_scanners_api.py [hunk 26]
```diff
@@ -69,3 +69,45 @@ def test_get_unknown_scanner(self, test_app):
...
+        )
+        assert response.status_code == BAD_REQUEST
+
+    def test_add_unknown_scanner_status(self, test_app):
+        response = test_app.put(
+            self.URI + "/42/status",
+            data=json.dumps({"job": "foo"}),
+            headers={'Content-Type': 'application/json'}
+        )
+        assert response.status_code == CREATED
```
Comment (local-minimum) [inline tests/unit/ui_server/test_scanners_api.py]: Missing test for adding a new scanner (`id`) requesting existing name of other scanner. Also missing test for adding a new scanner without having a suggested name and the generator produces a duplicate name (mock the generator).

**Comment by local-minimum** [comment]
Also, very importantly. This cannot be merged unless the js-UI is updated to request not only scanners but scanner statuses too.

File: scanomatic/io/scanning_store.py (hunk 23 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py (hunk 24 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: leaving for later

File: tests/unit/ui_server/test_scanners_api.py (hunk 26 above)
Comment (skymandr) [reply tests/unit/ui_server/test_scanners_api.py]: leaving for later (cascade from above requirement).

File: scanomatic/ui_server/serialization.py (hunk 25 above)
Comment (skymandr) [reply scanomatic/ui_server/serialization.py]: done

File: scanomatic/io/generic_name.py [hunk 27]
```diff
@@ -0,0 +1,40 @@
+from random import Random
+
```
Comment (gdetrez) [inline scanomatic/io/generic_name.py]: I'm not sure this belongs in io. Maybe `scanomatics.utils` would be more appropriate?

File: scanomatic/io/scanning_store.py [hunk 28]
```diff
@@ -22,22 +30,28 @@ def __init__(self):
...
+        self._scanners[scanner_id] = Scanner(name, scanner_id)
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: As before, this class shouldn't be responsible for creating new domain objects

File: scanomatic/io/generic_name.py (hunk 27 above)
Comment (skymandr) [reply scanomatic/io/generic_name.py]: I agree.

File: scanomatic/io/scanning_store.py (hunk 28 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: To be frank, I don't remember that discussion and I'm not quite sure what you mean by domain object... Care to clarify?

Comment (gdetrez) [reply scanomatic/io/scanning_store.py]: This module is responsible for storing/retrieving domain objects (like `Scanner`, `ScannerStatus`, etc) not creating them or deciding for default values. That's domain logic, which for now is mostly in the views (not ideal, I agree, but let's keep it consistent...) ([Discussion](https://github.com/Scan-o-Matic/scanomatic/pull/265#discussion_r163274043))

File: scanomatic/io/scanning_store.py [hunk 29]
```diff
@@ -127,3 +128,19 @@ def get_latest_scanner_status(self, scanner_id):
 
     def add_scanner_status(self, scanner_id, status):
         self.get_scanner_status_list(scanner_id).append(status)
+
+    def scanner_is_online(self, scanner_id):
```
Comment (gdetrez) [inline scanomatic/io/scanning_store.py]: This follows the same argument: deciding wether a scanner is online or not based on the latest status shouldn't be a responsibility of the storing module.

File: scanomatic/io/scanning_store.py (hunk 28 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: Ok. I seem to remember that @local-minimum had a different opinion on this, though I'm sure he could be convinced. I would like to know what he thinks before changing anything though.

File: scanomatic/io/scanning_store.py (hunk 29 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: So implement this in a view in `scanners_api.py` instead?

Comment (gdetrez) [reply scanomatic/io/scanning_store.py]: Not necessarily. You can also update the existing view `GET /api/scanners/<id>` to compute the online status on the fly.

Comment (skymandr) [reply scanomatic/io/scanning_store.py]: Seems @local-minimum agrees. I'll make it so.

File: scanomatic/io/scanning_store.py (hunk 28 above)
Comment (local-minimum) [reply scanomatic/io/scanning_store.py]: I noticed that @gdetrez placed similar type declaration in `scanomatic/models` and that feels consistent with the rest of SoM. I trust @gdetrez knows better on how to best divide those things so do as he says :smiley_cat:

File: scanomatic/io/scanning_store.py (hunk 29 above)
Comment (local-minimum) [reply scanomatic/io/scanning_store.py]: Do as he says :+1:, that was kind of what my edited comment wanted to say.

File: scanomatic/io/scanning_store.py (hunk 24 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done

File: scanomatic/io/scanning_store.py (hunk 28 above)
Comment (skymandr) [reply scanomatic/io/scanning_store.py]: done, I think

File: scanomatic/ui_server/serialization.py [hunk 30]
```diff
@@ -16,9 +16,11 @@ def job2json(job):
     return obj
 
 
-def status2json(status):
+def status2json(status, power=False, owner=None):
```
Comment (gdetrez) [inline scanomatic/ui_server/serialization.py]: I think to respect the existing API, `power` and `owner` should be added to scanners, not statuses.

Comment (skymandr) [reply scanomatic/ui_server/serialization.py]: I'm beginning to feel that we're going around in circles here...

Comment (skymandr) [reply scanomatic/ui_server/serialization.py]: done

File: scanomatic/ui_server/scanners_api.py [hunk 31]
```diff
@@ -21,14 +35,68 @@ def scanners_get():
         scanning_store.get_free_scanners() if get_free else
         scanning_store.get_all_scanners()
     )
-    return jsonify([scanner._asdict() for scanner in scanners])
+    return jsonify([
+        scanner2json(
+            scanner,
+            _scanner_is_online(scanner.identifier, scanning_store)
+        ) for scanner in scanners
```
Comment (local-minimum) [inline scanomatic/ui_server/scanners_api.py]: Missing `owner` which is the concept of the scanner being busy with a job (should get better name in near future).  My idea was to have it be missing if there's no job and just have the serialized job there if there's a job running.

File: scanomatic/ui_server/serialization.py [hunk 32]
```diff
@@ -14,3 +14,20 @@ def job2json(job):
...
+    if status.server_time is not None:
+        assert is_utc(status.server_time)
+        obj['serverTime'] = status.server_time.strftime('%Y-%m-%dT%H:%M:%SZ')
+    return obj
+
+
+def scanner2json(scanner, power=False, owner=None):
+    obj = scanner._asdict()
+    obj['power'] = power
+    obj['owner'] = owner
```
Comment (local-minimum) [inline scanomatic/ui_server/serialization.py]: Should only insert owner if there is a job and omit if not (like the other functions above).
Should use `job2json` if there's a job/owner.

File: scanomatic/ui_server/serialization.py [hunk 33]
```diff
@@ -14,3 +14,20 @@ def job2json(job):
         assert is_utc(job.start_time)
         obj['startTime'] = job.start_time.strftime('%Y-%m-%dT%H:%M:%SZ')
     return obj
+
+
+def status2json(status):
```
Comment (local-minimum) [inline scanomatic/ui_server/serialization.py]: rename to `scanner_status2json`

File: scanomatic/ui_server/serialization.py [hunk 34]
```diff
@@ -14,3 +14,20 @@ def job2json(job):
...
+    obj = {
+        'job': status.job,
```
Comment (local-minimum) [inline scanomatic/ui_server/serialization.py]: Should not insert `'job'` if `status.job` is `None`.

File: tests/unit/io/test_scanning_store.py [hunk 35]
```diff
@@ -33,40 +33,36 @@ def scanning_store():
...
 
-    @pytest.mark.parametrize('scanner', (SCANNER, SCANNER_POWER))
+    @pytest.mark.parametrize('scanner', (SCANNER_ONE, SCANNER_TWO))
     def test_getting_scanner(self, scanning_store, scanner):
         assert scanning_store.get_scanner(scanner.identifier) == scanner
 
     def test_get_free(self, scanning_store):
         assert set(scanning_store.get_free_scanners()) == {
-            SCANNER, SCANNER_POWER,
+            SCANNER_ONE, SCANNER_TWO,
```
Comment (local-minimum) [inline tests/unit/io/test_scanning_store.py]: Now that a scanner can be non-free, the test should be made so that it ensures that occupied scanners are omitted when requesting free scanners.

File: scanomatic/ui_server/scanners_api.py [hunk 36]
```diff
@@ -21,14 +35,68 @@ def scanners_get():
         scanning_store.get_free_scanners() if get_free else
         scanning_store.get_all_scanners()
     )
-    return jsonify([scanner._asdict() for scanner in scanners])
+    return jsonify([
+        scanner2json(
+            scanner,
+            _scanner_is_online(scanner.identifier, scanning_store)
```
Comment (local-minimum) [inline scanomatic/ui_server/scanners_api.py]: Missing test that checks the behaviour of `_scanner_is_online`

File: scanomatic/ui_server/scanners_api.py (hunk 31 above)
Comment (gdetrez) [reply scanomatic/ui_server/scanners_api.py]: @local-minimum: I suggest to skip that: it duplicates the result of the `/scanner/<id>/job` endpoint, which would mean duplicating the tests, etc... The ui can use this endpoint instead maybe?

Comment (local-minimum) [reply scanomatic/ui_server/scanners_api.py]: OK, lets just not say anything about owners in this PR, but then remove from the `scanner2json` serializer too. The UI will tell some lies, but we'll make a new issue for next spring.

File: tests/unit/io/test_scanning_store.py (hunk 35 above)
Comment (gdetrez) [reply tests/unit/io/test_scanning_store.py]: Since no one uses the feature, I suggest to skip that and remove the `get_free_scanners` method and corresponding endpoint code. I can help with that if you don't have time.
